copy/copy.go

```diff
@@ -43,7 +43,9 @@ type imageCopier struct {
 	rawSource         types.ImageSource
 	diffIDsAreNeeded  bool
 	canModifyManifest bool
-	reportWriter      io.Writer
+	writer            io.Writer
+	reportWriter      func(f string, a ...interface{})
+	hasReportWriter   bool
```

Comment:
It’s not immediately clear what is the difference between `writer`/`reportWriter` here. The naming is a bit confusing, as is `hasReportWriter` for a `reportWriter` which always exists

And if this is moving all all users the function, and the `io.Writer` is used _only_ for initializing the progress bar, we don’t really need `hasReportWriter`; we could just have an `io.Writer` which is either an implementation or `nil`.